File: apps/api/core/api/routes/live.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""Public, privacy-preserving projection of SeaCommons live signals."""

from __future__ import annotations

import asyncio
import hashlib
import json
import re
from datetime import datetime, timezone
from typing import Optional

from fastapi import APIRouter, HTTPException, Query, Request, WebSocket, WebSocketDisconnect

from core.config import config
from core.intel.store import intel_store
from core.live.feed import public_drift_collection, public_signal_collection
from core.live.projection import _public_intel_feature

router = APIRouter(prefix="/api/v1/live", tags=["live"])


@router.get("/signals")
async def live_signals(
    limit: int = Query(300, ge=1, le=500),
    days: int = Query(30, ge=1, le=365),
    since: Optional[str] = Query(None),
    mode: str = Query(
        "humanitarian",
        pattern="^(humanitarian|maritime|security|all)$",
        description=(
            "humanitarian: public-eligible SAR/humanitarian output. "
            "maritime: public-eligible Safety plus reviewed/published Maritime output. "
            "security is a temporary alias for maritime. all: both compartments."
        ),
    ),
):
    """Public map-ready signal feed. No private inbound content is returned."""
    return public_signal_collection(limit=limit, days=days, since=since, mode=mode)


@router.get("/signals/{event_id}/response")
async def live_signal_response(event_id: str, request: Request):
    """
    On-demand NGO cross-check for a single live episode.

    Returns which known SAR NGO / coastguard vessels are within range and
    heading toward the episode (bearing/ETA/heading), when each track was last
    saved to the AIS registry, recent motion flags (speed spike, search
    pattern, sudden stop, loitering, rescue cluster — reused from the AIS
    spike detector's stored observations), and other live signals nearby
    (cross-check). Also returns a GeoJSON FeatureCollection (`geojson`) of the
    episode→vessel lines and vessel points for direct map rendering.

    Privacy: only public AIS telemetry and the already-public signal metadata
    are returned — never raw messages or identifiers. Anonymous + rate limited
    like the other live read endpoints.
    """
    from core.api.ratelimit import rate_limit
    from core.intel.ngo_response import analyze_ngo_response

    rate_limit(request, max_per_minute=30, scope="live-response")
    normalized = event_id.removeprefix("intel:")
    event = intel_store.get(normalized)
    if event is None:
        raise HTTPException(status_code=404, detail="Signal not found")
    if event.lat is None or event.lon is None:
        raise HTTPException(status_code=422, detail="Signal has no position")
    if _public_intel_feature(event) is None:
        raise HTTPException(status_code=404, detail="Signal not found")

    related_signals = [
        other
        for other in intel_store.events(limit=300, max_age_days=3)
        if _public_intel_feature(other) is not None
    ]
    try:
        return analyze_ngo_response(event, related_signals=related_signals)
    except ValueError:
        raise HTTPException(status_code=422, detail="Signal has no position")


@router.get("/vessels/{mmsi}/context")
async def live_vessel_context(
    mmsi: str,
    request: Request,
    hours: float = Query(default=168.0, ge=1, le=24 * 30),
):
    """Public-AIS identity, observed track and sanctions-list explanation."""
    from core.api.ratelimit import rate_limit
    from core.api.routes.mda import build_vessel_dossier

    rate_limit(request, max_per_minute=30, scope="live-vessel-context")
    if re.fullmatch(r"\d{9}", mmsi) is None:
        raise HTTPException(status_code=422, detail="MMSI must contain exactly 9 digits")
    return build_vessel_dossier(mmsi, hours=hours, track_limit=240)


@router.get("/drifts")
async def live_drifts(limit: int = Query(100, ge=1, le=200)):
    """Public map-ready drift products, kept separate from received signals."""
    return public_drift_collection(limit=limit)


@router.get("/hypotheses")
async def live_hypotheses(limit: int = Query(100, ge=1, le=200)):
    """Published Maritime Intelligence hypotheses (docs/fixes.md M14.3/
    M14.4) -- core.intel.publication_policy.project_public_maritime_
    assessed() is the sole authority for what this returns; a hypothesis
    still in candidate/collecting/review_ready/assessed never appears
    here."""
    from core.intel.hypothesis_publication import public_hypothesis_collection

    return public_hypothesis_collection(limit=limit)


@router.get("/archives")
async def live_archives(limit: int = Query(40, ge=1, le=200)):
    """Anonymised incident index for the Play archive timeline.

    Two kinds of archived incident carry a computed drift: SAR cases opened
    through /api/v1/alert, and OSINT distress events (Alarm Phone, ...) whose
    auto-drift completed. Both are surfaced here, newest first, with only the
    coarse fields Play needs — never the source message or any identifier.
    """
    from core.db.models import DriftResultDB, IntelEventDB
    from core.db.session import session_scope
    from core.db.store import list_alerts

    seen: set[str] = set()
    archives: list[dict] = []

    for alert in list_alerts(limit=limit * 3):
        if alert.get("status") != "completed":
            continue
        event = alert.get("event") or {}
        archives.append({
            "id": alert["event_id"],
            "timestamp": event.get("timestamp"),
            "lat": event.get("lat"),
            "lon": event.get("lon"),
            "vessel_type": event.get("vessel_type") or "case",
            "persons": event.get("persons") or 1,
            "kind": "sar_case",
        })
        seen.add(alert["event_id"])

    try:
        from sqlalchemy import func, literal, or_

        with session_scope() as db:
            # An intel auto-drift persists drift_results.event_id as
            # "intel:<id>" (see core.intel.drift_service); the seeded SAR
            # archives use the bare id. Match both.
            rows = (
                db.query(DriftResultDB, IntelEventDB)
                .join(
                    IntelEventDB,
                    or_(
                        DriftResultDB.event_id == IntelEventDB.id,
                        DriftResultDB.event_id == literal("intel:") + IntelEventDB.id,
                    ),
                )
                .filter(DriftResultDB.status == "completed")
                .filter(IntelEventDB.lat.isnot(None))
                .order_by(func.coalesce(DriftResultDB.created_at, IntelEventDB.created_at).desc())
                .limit(limit * 4)
                .all()
            )
            for drift, ev in rows:
                if ev.id in seen:
                    continue
                meta = ev.meta or {}
                if not (meta.get("is_distress") or ev.type in {"distress", "vessel_incident"}):
                    continue
                archives.append({
                    "id": ev.id,
                    "timestamp": ev.timestamp_utc,
                    "lat": ev.lat,
                    "lon": ev.lon,
                    "vessel_type": meta.get("vessel_type") or "rubber_boat",
                    "persons": meta.get("persons") or meta.get("people") or 1,
                    "kind": "osint_distress",
                })
                seen.add(ev.id)
    except Exception:  # pragma: no cover - archive listing is best-effort
        pass

    archives.sort(key=lambda a: str(a.get("timestamp") or ""), reverse=True)
    return {
        "archives": archives[:limit],
        "generated_at": datetime.now(timezone.utc).isoformat(),
    }


@router.get("/archives/{event_id}/geojson")
async def live_archive_geojson(event_id: str):
    """Derived geometry for a public Play archive; no source message or identity."""
    from fastapi import HTTPException

    from core.db.store import get_alert, get_drift

    alert = get_alert(event_id)
    drift = get_drift(event_id)  # seeded SAR archives: drift_id == event_id
    if drift is None:
        # An intel auto-drift row is keyed by job id; its event_id is
        # "intel:<id>" (or the bare id). Take the newest completed one.
        from core.db.models import DriftResultDB
        from core.db.session import session_scope
        from core.db.store import drift_to_dict
        from sqlalchemy import func, or_

        with session_scope() as db:
            row = (
                db.query(DriftResultDB)
                .filter(
                    DriftResultDB.status == "completed",
                    or_(
                        DriftResultDB.event_id == event_id,
                        DriftResultDB.event_id == f"intel:{event_id}",
                    ),
                )
                .order_by(func.coalesce(DriftResultDB.created_at, None).desc())
                .first()
            )
            drift = drift_to_dict(row)

    alert_ok = alert is not None and alert.get("status") == "completed"
    drift_ok = bool(drift) and drift.get("status") == "completed"
    if not drift_ok or (alert is not None and not alert_ok):
        raise HTTPException(status_code=404, detail="Archive not found")
    features = [
        feature
        for feature in (
            drift.get("trajectory"),
            drift.get("cone_6h"),
            drift.get("cone_12h"),
            drift.get("cone_24h"),
        )
        if feature
    ]
    features.extend((drift.get("impact_point") or {}).get("features", []))
    return {"type": "FeatureCollection", "features": features}


@router.get("/pipeline")
async def live_pipeline():
    """Public-safe health for the single SeaCommons acquisition pipeline."""
    from core.acquisition.status import (
        acquisition_status_sources,
        ensure_default_acquisition_status,
    )

    ensure_default_acquisition_status()
    return {
        "generated_at": datetime.now(timezone.utc).isoformat(),
        "sources": acquisition_status_sources(),
    }


@router.get("/sources")
async def live_sources():
    """Public health summary without credentials, endpoint URLs or raw errors."""
    from core.intel.source_registry import source_registry

    registry_sources = {
        source["name"]: source
        for source in source_registry.get_all()
        if source["name"]
        in {
            "X / Twitter",
            "X / Twitter (twikit)",
            "Mastodon",
            "Official NGO RSS",
            "GDACS",
            "Bluesky",
        }
    }
    try:
        from core.connectors.service import status_counts
        from core.db.session import session_scope

        with session_scope() as db:
            whatsapp_connectors = status_counts(db, "whatsapp_cloud")
    except Exception:
        whatsapp_connectors = {}
    whatsapp_ready = bool(
        config.META_APP_ID
        and config.META_APP_SECRET
        and config.META_WEBHOOK_VERIFY_TOKEN
        and whatsapp_connectors.get("active", 0)
    )
    expected = (
        ("X / Twitter (twikit)", "twitter", bool(config.TWIKIT_ENABLED)),
        ("X / Twitter", "twitter", bool(config.TWITTER_BEARER_TOKEN)),
        ("WhatsApp partner intake", "whatsapp", whatsapp_ready),
        (
            "Telegram intake",
            "telegram",
            bool(config.TELEGRAM_BOT_TOKEN and config.TELEGRAM_WEBHOOK_SECRET),
        ),
        ("Partner webhook", "partner", bool(config.PARTNER_WEBHOOK_SECRET)),
    )
    sources = []
    for name, source_type, configured in expected:
        observed = registry_sources.pop(name, None)
        sources.append(
            {
                "name": name,
                "type": source_type,
                "status": observed["status"]
                if observed
                else ("pending" if configured else "offline"),
                "last_poll_at": observed["last_poll_at"] if observed else None,
                "events_last_hour": observed["events_last_hour"] if observed else 0,
                "total_events": observed["total_events"] if observed else 0,
                "consecutive_errors": observed["consecutive_errors"] if observed else 0,
                "pipeline_status": observed.get("pipeline_status", observed["status"])
                if observed
                else ("pending" if configured else "offline"),
                "source_status": observed.get("source_status", "unknown")
                if observed
                else "unknown",
                "configured": observed.get("configured", int(configured))
                if observed
                else int(configured),
                "reachable": observed.get("reachable", 0) if observed else 0,
                "handles": [
                    {
                        "name": handle["name"],
                        "status": handle["status"],
                        "last_poll_at": handle["last_poll_at"],
                        "total_events": handle["total_events"],
                    }
                    for handle in observed.get("handles", [])
                ]
                if observed
                else [],
            }
        )
    for observed in registry_sources.values():
        sources.append(
            {
                "name": observed["name"],
                "type": observed["type"],
                "status": observed["status"],
                "last_poll_at": observed["last_poll_at"],
                "events_last_hour": observed["events_last_hour"],
                "total_events": observed["total_events"],
                "consecutive_errors": observed["consecutive_errors"],
                "pipeline_status": observed.get("pipeline_status", observed["status"]),
                "source_status": observed.get("source_status", "unknown"),
                "configured": observed.get("configured", 0),
                "reachable": observed.get("reachable", 0),
                "handles": [
                    {
                        "name": handle["name"],
                        "status": handle["status"],
                        "last_poll_at": handle["last_poll_at"],
                        "total_events": handle["total_events"],
                    }
                    for handle in observed.get("handles", [])
                ],
            }
        )
    active = sum(1 for source in sources if source["status"] == "active")
    return {
        "sources": sources,
        "summary": {
            "total": len(sources),
            "active": active,
            "degraded": sum(1 for source in sources if source["status"] == "degraded"),
            "offline": sum(1 for source in sources if source["status"] == "offline"),
            "pending": sum(1 for source in sources if source["status"] == "pending"),
        },
        "channels": {
            "twitter": bool(config.TWITTER_BEARER_TOKEN),
            "twitter_alarm_phone": any(
                source["name"] == "X / Twitter (twikit)"
                and source["status"] in {"active", "degraded"}
                and source["reachable"] > 0
                for source in sources
            ),
            "whatsapp": whatsapp_ready,
            "whatsapp_active_connectors": whatsapp_connectors.get("active", 0),
            "telegram": bool(config.TELEGRAM_BOT_TOKEN and config.TELEGRAM_WEBHOOK_SECRET),
            "partner_webhook": bool(config.PARTNER_WEBHOOK_SECRET),
        },
        "collector": {
            "mode": "continuous",
            "browser_independent": True,
            "persistence": "database",
            "supervisor": "systemd",
        },
        "generated_at": datetime.now(timezone.utc).isoformat(),
    }


@router.get("/ngo-vessels")
async def live_ngo_vessels():
    """Public projection of known SAR NGO/coastguard vessel positions.

    Same data and query as the authenticated operator route
    (/api/v1/intel/ngo) — these vessels already broadcast AIS publicly, so
    there is nothing to withhold; this just gives the public Live map a
    route under the same public /api/v1/live/ prefix as everything else it
    reads, instead of poking a hole in /api/v1/intel's auth gate.
    """
    from core.intel.ngo_registry import ngo_vessel_geojson

    return ngo_vessel_geojson()


@router.get("/platforms")
async def live_platforms():
    """Public projection of Mediterranean oil/gas platform positions
    (static public data — EMODnet/OGA/OSM). Mirrors /api/v1/zones/platforms,
    which is already unauthenticated, under the public Live namespace."""
    from core.api.routes.zones import get_platforms

    return await get_platforms()


@router.websocket("/stream")
async def live_stream(websocket: WebSocket, mode: str = "humanitarian"):
    """Public WebSocket snapshot stream; the browser falls back to REST polling."""
    selected_mode = mode if mode in {"humanitarian", "security", "all"} else "humanitarian"
    await websocket.accept()
    previous_digest = ""
    try:
        while True:
            snapshot = public_signal_collection(limit=500, days=30, mode=selected_mode)
            payload = json.dumps(
                {
                    "type": "snapshot",
                    "features": snapshot["features"],
                    "meta": snapshot["meta"],
                },
                separators=(",", ":"),
            )
            digest = hashlib.blake2s(payload.encode(), digest_size=8).hexdigest()
            if digest != previous_digest:
                await websocket.send_text(payload)
                previous_digest = digest
            else:
                await websocket.send_text('{"type":"ping"}')
            await asyncio.sleep(10)
    except WebSocketDisconnect:
        return
    except Exception:
        try:
            await websocket.close()
        except Exception:
            pass

File: apps/web/api/live.js
import http from 'node:http';

const UPSTREAM_HOST = '204.216.210.155';
const UPSTREAM_VIRTUAL_HOST = 'api.seacommons.org';

function requestJson(path) {
  return new Promise((resolve, reject) => {
    const request = http.get(
      {
        hostname: UPSTREAM_HOST,
        port: 80,
        path,
        headers: {
          host: UPSTREAM_VIRTUAL_HOST,
          accept: 'application/json',
          'user-agent': 'SeaCommons-Vercel-Live/1.0',
        },
      },
      (response) => {
        const chunks = [];
        response.on('data', (chunk) => chunks.push(chunk));
        response.on('end', () => {
          const body = Buffer.concat(chunks).toString('utf8');
          let data = null;
          try { data = body ? JSON.parse(body) : null; } catch { /* handled below */ }
          resolve({ status: response.statusCode || 502, data });
        });
      },
    );
    request.setTimeout(24_000, () => request.destroy(new Error('SeaCommons API timed out')));
    request.on('error', reject);
  });
}

function alertFallback() {
  const features = [];
  return {
    type: 'FeatureCollection',
    features,
    meta: {
      schema: 'org.seacommons.live-feed/v1',
      total: features.length,
      with_coords: features.length,
      generated_at: new Date().toISOString(),
      compatibility_mode: true,
      privacy: 'published signals only; private identifiers and raw messages excluded',
    },
  };
}

function driftFallback(payload) {
  const features = (payload?.features || [])
    .filter((feature) => ['LineString', 'Polygon', 'Point'].includes(feature?.geometry?.type))
    .map((feature) => ({
      type: 'Feature',
      geometry: feature.geometry,
      properties: {
        type: feature.properties?.type || null,
        intel_severity: 'medium',
        intel_source: 'SeaCommons engine',
        auto_drift: true,
        compatibility_mode: true,
      },
    }));
  return {
    type: 'FeatureCollection',
    features,
    meta: {
      schema: 'org.seacommons.live-drift/v1',
      drifts: features.filter((feature) => feature.geometry.type === 'LineString').length,
      generated_at: new Date().toISOString(),
      compatibility_mode: true,
    },
  };
}

function emptyDriftFallback() {
  return {
    type: 'FeatureCollection',
    features: [],
    meta: {
      schema: 'org.seacommons.live-drift/v1',
      drifts: 0,
      generated_at: new Date().toISOString(),
      compatibility_mode: true,
      visibility: 'play_only',
    },
  };
}

function archiveFallback(payload) {
  const archives = (Array.isArray(payload) ? payload : [])
    .filter((alert) => alert?.status === 'completed')
    .slice(0, 8)
    .map((alert) => ({
      id: alert.event_id,
      timestamp: alert.event?.timestamp || null,
      lat: Number(alert.event?.lat),
      lon: Number(alert.event?.lon),
      vessel_type: alert.event?.vessel_type || 'case',
      persons: Number(alert.event?.persons || 1),
    }));
  return { archives, generated_at: new Date().toISOString(), compatibility_mode: true };
}

function sourceFallback(summary) {
  const configured = {
    twitter: Boolean(summary?.channels?.twitter?.configured),
    whatsapp: Boolean(summary?.channels?.whatsapp?.configured),
    telegram: Boolean(summary?.channels?.telegram?.configured),
    partner: Boolean(summary?.channels?.partner_webhook?.configured),
  };
  const definitions = [
    ['X / Twitter', 'twitter', configured.twitter],
    ['WhatsApp intake', 'whatsapp', configured.whatsapp],
    ['Telegram intake', 'telegram', configured.telegram],
    ['Partner webhook', 'partner', configured.partner],
  ];
  const sources = definitions.map(([name, type, active]) => ({
    name,
    type,
    status: active ? 'active' : 'offline',
    last_poll_at: active ? summary?.generated_at || null : null,
    events_last_hour: 0,
    total_events: 0,
    consecutive_errors: 0,
  }));
  const active = sources.filter((source) => source.status === 'active').length;
  return {
    sources,
    summary: {
      total: sources.length,
      active,
      degraded: 0,
      offline: sources.length - active,
    },
    channels: {
      twitter: configured.twitter,
      whatsapp: configured.whatsapp,
      telegram: configured.telegram,
      partner_webhook: configured.partner,
    },
    collector: {
      mode: 'continuous',
      browser_independent: true,
      persistence: 'server',
      supervisor: 'systemd',
    },
    generated_at: new Date().toISOString(),
    compatibility_mode: true,
  };
}

export default async function handler(req, res) {
  res.setHeader('cache-control', 'no-store');
  res.setHeader('content-type', 'application/json; charset=utf-8');
  res.setHeader('x-seacommons-live-version', '2026-08-01.2');
  if (req.method !== 'GET' && req.method !== 'HEAD') {
    res.setHeader('allow', 'GET, HEAD');
    return res.status(405).json({ detail: 'Method not allowed' });
  }

  const resource = Array.isArray(req.query.resource) ? req.query.resource[0] : req.query.resource;
  const eventId = Array.isArray(req.query.event_id) ? req.query.event_id[0] : req.query.event_id;
  const safeEventId = typeof eventId === 'string' && /^[a-zA-Z0-9-]{6,80}$/.test(eventId)
    ? eventId
    : '';
  const requestedLimit = Math.min(500, Math.max(1, Number.parseInt(req.query.limit, 10) || 500));
  const requestedDays = Math.min(365, Math.max(1, Number.parseInt(req.query.days, 10) || 30));
  const upstreamPath = resource === 'sources'
    ? '/api/v1/live/sources'
    : resource === 'drifts'
      ? '/api/v1/live/drifts?limit=100'
      : resource === 'archives'
        ? '/api/v1/live/archives?limit=8'
        : resource === 'archive' && safeEventId
          ? `/api/v1/live/archives/${safeEventId}/geojson`
          : `/api/v1/live/signals?limit=${requestedLimit}&days=${requestedDays}`;
  try {
    const upstream = await requestJson(upstreamPath);
    if (upstream.status === 200 && upstream.data) {
      res.setHeader('x-seacommons-live-source', 'engine');
      return res.status(200).json(upstream.data);
    }

    let fallback;
    if (resource === 'sources') {
      fallback = sourceFallback((await requestJson('/api/v1/ops/summary')).data);
    } else if (resource === 'drifts') {
      fallback = emptyDriftFallback();
    } else if (resource === 'archives') {
      fallback = archiveFallback((await requestJson('/api/v1/alerts')).data);
    } else if (resource === 'archive' && safeEventId) {
      fallback = driftFallback((await requestJson(`/api/v1/alert/${safeEventId}/geojson`)).data);
    } else {
      fallback = alertFallback();
    }
    res.setHeader('x-seacommons-live-source', 'compatibility');
    return res.status(200).json(fallback);
  } catch {
    return res.status(502).json({ detail: 'SeaCommons live feed unavailable' });
  }
}

export const config = {
  maxDuration: 30,
};
